### PR TITLE
Distinguish "missing repository" from "missing repository plugin"

### LINKED ISSUE #81758: Distinguish "repository doesn't exist" from "repository plugin not installed"
**Elasticsearch version** (`bin/elasticsearch --version`): 7.16.1 (and many others)

**Plugins installed**: []

**JVM version** (`java -version`): Any

**OS version** (`uname -a` if on a Unix-like system): Any

**Description of the problem including expected versus actual behavior**:

If you define a repository using a plugin and then remove the plugin then the `RepositoryMetadata` lives on but we do not create a corresponding `Repository` instance. This means that operations on the repository (creating snapshots etc) will fail with a `RepositoryMissingException` which is very confusing because `GET _snapshot` will show that the repository is not missing at all.

I believe in this case we should create a placeholder `Repository` instance so that operations on the repository fail with a more helpful message.

### PR DESCRIPTION
**Pull request by idegtiarenko**
Currently it is possible to uninstall repository plugin before
deleting repository that uses it. In such case the repository is still
listed in the API however every interraction with it produces the
"no such repository" exception. This is very confusing for the users.
This change updates the error message to correctly point the root of the
 issue.

Closes: #81758

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/MissingPluginRepository.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    @Override
+    public void verify(String verificationToken, DiscoveryNode localNode) {
+        throw createMissingPluginException();
+    }
+
+    @Override
+    public boolean isReadOnly() {
+        return true;
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/repositories/MissingPluginRepository.java:116]: Currently this result in:
```
PUT http://localhost:9200/_snapshot/repository-hdfs/test

>>>
{
  "error": {
    "root_cause": [
      {
        "type": "repository_exception",
        "reason": "[repository-hdfs] cannot create snapshot in a readonly repository"
      }
    ],
    "type": "repository_exception",
    "reason": "[repository-hdfs] cannot create snapshot in a readonly repository"
  },
  "status": 500
}
```

I do not think this is correct and looking for a way to return no plugin exception when taking a snapshot.

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java [hunk 2]
```diff
@@ -651,7 +657,10 @@ private Repository createRepository(RepositoryMetadata repositoryMetadata, Map<S
         logger.debug("creating repository [{}][{}]", repositoryMetadata.type(), repositoryMetadata.name());
         Repository.Factory factory = factories.get(repositoryMetadata.type());
         if (factory == null) {
-            throw new RepositoryException(repositoryMetadata.name(), "repository type [" + repositoryMetadata.type() + "] does not exist");
+            throw new RepositoryPluginException(
+                repositoryMetadata.name(),
+                "repository type [" + repositoryMetadata.type() + "] does not exist"
+            );
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java:663]:
```
[2022-01-12T09:22:56,827][WARN ][o.e.r.RepositoriesService] [idegtiarenko-ThinkPad-T490] failed to create repository [repository-hdfs]
org.elasticsearch.repositories.RepositoryPluginException: [repository-hdfs] repository type [hdfs] does not exist
	at org.elasticsearch.repositories.RepositoriesService.createRepository(RepositoriesService.java:662) ~[elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at org.elasticsearch.repositories.RepositoriesService.applyClusterState(RepositoriesService.java:505) [elasticsearch-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
```

Should I reword this to `repository plugin for type [type] does not exist` or keep existing message?

File: server/src/main/java/org/elasticsearch/repositories/MissingPluginRepository.java (hunk 1 above)
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/repositories/MissingPluginRepository.java]: Resolved: 

```
http://localhost:9200/_snapshot/repository-hdfs/test

>>>
HTTP/1.1 500 Internal Server Error
X-elastic-product: Elasticsearch
content-type: application/json;charset=utf-8

{
  "error": {
    "root_cause": [
      {
        "type": "repository_plugin_exception",
        "reason": "[repository-hdfs] the plugin is not installed for repository type [hdfs]"
      }
    ],
    "type": "repository_plugin_exception",
    "reason": "[repository-hdfs] the plugin is not installed for repository type [hdfs]"
  },
  "status": 500
}
```

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java:663]: Blaming a missing plugin here seems wrong, it could just be a typo, but if `createRepository` knew whether to fail like this or not then I'd suggest the same message as I suggested above:

`"repository type [" + repositoryMetadata.type() + "] is unknown; ensure that all required plugins are installed on this node"`

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java [hunk 3]
```diff
@@ -162,6 +164,44 @@ public void testRepositoriesStatsCanHaveTheSameNameAndDifferentTypeOverTime() {
...
 
+    // this can happen when the repository plugin is removed, but repository is still exist
+    public void testHandlesUnknownRepositoryTypeWhenApplyingClusterState() {
+        var repoName = randomAlphaOfLengthBetween(10, 25);
+
+        var clusterState = createClusterStateWithRepo(repoName, "unknown");
+        repositoriesService.applyClusterState(new ClusterChangedEvent("starting", clusterState, emptyState()));
+
+        var repo = repositoriesService.repository(repoName);
+        assertThat(repo, isA(MissingPluginRepository.class));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java:175]: Given that this is a user-facing exception, I like to have some assertions about the message itself rather than just the exception type.

**Review by DaveCTurner — COMMENTED** [review]
Looks good @idegtiarenko, I left a few comments/suggestions.

File: server/src/main/java/org/elasticsearch/repositories/MissingPluginRepository.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * This class represents a repository that still exists in the global cluster state but could not be loaded as its plugin was uninstalled.
+ */
+public class MissingPluginRepository extends AbstractLifecycleComponent implements Repository {
```
[the remark this answers is not in the document]
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/repositories/MissingPluginRepository.java:34]: I was struggling to come up with a good name. Will rename!

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java [hunk 5]
```diff
@@ -162,6 +165,61 @@ public void testRepositoriesStatsCanHaveTheSameNameAndDifferentTypeOverTime() {
...
+            code.run();
+        } catch (Throwable e) {
+            assertThat(e, exceptionType, messageMatcher);
+        }
+    }
+
+    private static void assertThat(Throwable exception, Class<? extends Exception> exceptionType, Matcher<String> messageMatcher) {
+        assertThat(exception, isA(exceptionType));
+        assertThat(exception.getMessage(), messageMatcher);
+    }
```
Comment (idegtiarenko) [inline server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java:221]: Is there a good common place to move this to? May be ESTestCase?
I am missing a simple way to assert exceptions in hamcrest similarly to what assertj offers: https://assertj.github.io/doc/#assertj-core-exception-assertions

**Comment by idegtiarenko** [comment]
@elasticmachine run elasticsearch-ci/rest-compatibility

File: server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java (hunk 5 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceTests.java:221]: TBH we mostly don't care about the exception type, the message is the important bit. I think it'd fit somewhere in `:test:framework:org.elasticsearch.test.hamcrest` if you wanted to add a utility method for this tho.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by idegtiarenko** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Review by kirakoki — APPROVED** [review]
